boa_gc/src/lib.rs

```diff
@@ -1,6 +1,448 @@
+                let element_pointer = Box::into_raw(Box::from(gc_box));
+
+                let old_start = gc.adult_start.take();
+                (*element_pointer).set_header_pointer(old_start);
+                (*element_pointer).value().unroot();
+
+                gc.adult_start
+                    .set(Some(NonNull::new_unchecked(element_pointer)));
+
+                gc.runtime.total_bytes_allocated += element_size;
+
+                WeakGc::new(NonNull::new_unchecked(element_pointer))
+            }
+        })
+    }
+
+    unsafe fn manage_state(gc: &mut BoaGc) {
+        if gc.runtime.adult_bytes > gc.config.adult_threshold {
+            Collector::run_full_collection(gc);
+
+            if gc.runtime.adult_bytes as f64
+                > gc.config.adult_threshold as f64 * gc.config.growth_ratio
+            {
+                gc.config.adult_threshold =
+                    (gc.runtime.adult_bytes as f64 / gc.config.growth_ratio) as usize
+            }
+        }
+    }
+}
+
+// This collector currently functions in four main phases
+//
+// Mark -> Finalize -> Mark -> Sweep
+//
+// Mark nodes as reachable then finalize the unreachable nodes. A remark phase
+// then needs to be retriggered as finalization can potentially resurrect dead
+// nodes.
+//
+// A better appraoch in a more concurrent structure may be to reorder.
+//
+// Mark -> Sweep -> Finalize
+pub struct Collector;
+
+impl Collector {
+    pub(crate) unsafe fn run_full_collection(gc: &mut BoaGc) {
+        let _timer = Profiler::global().start_event("Gc Full Collection", "gc");
+        gc.runtime.collections += 1;
+        let unreachable_adults = Self::mark_heap(&gc.adult_start);
+
+        // Check if any unreachable nodes were found and finalize
+        if !unreachable_adults.is_empty() {
+            Self::finalize(unreachable_adults);
+        }
+
+        let _final_unreachable_adults = Self::mark_heap(&gc.adult_start);
+
+        // Sweep both without promoting any values
+        Self::sweep(
+            &gc.adult_start,
+            &mut gc.runtime.adult_bytes,
+            &mut gc.runtime.total_bytes_allocated,
+        );
+    }
+
+    pub(crate) unsafe fn mark_heap(
+        head: &StdCell<Option<NonNull<GcBox<dyn Trace>>>>,
+    ) -> Vec<NonNull<GcBox<dyn Trace>>> {
+        let _timer = Profiler::global().start_event("Gc Marking", "gc");
+        // Walk the list, tracing and marking the nodes
+        let mut finalize = Vec::new();
+        let mut ephemeron_queue = Vec::new();
+        let mut mark_head = head;
+        while let Some(node) = mark_head.get() {
+            if (*node.as_ptr()).header.is_ephemeron() {
+                ephemeron_queue.push(node);
+            } else if (*node.as_ptr()).header.roots() > 0 {
+                (*node.as_ptr()).trace_inner();
+            } else {
+                finalize.push(node)
+            }
+            mark_head = &(*node.as_ptr()).header.next;
```

Comment:
Now that I think about it, the previous design probably didn't work because this only marked the nodes that were children of young nodes. This left unmarked all children of only adult nodes, causing erroneous drops.

Reply:
Yeah, I thought that too so I started running mark on the adult heap too, but even with that the nodes were still dropping. Which now that I type that out makes me think that `trace_inner` was the issue.